### PR TITLE
Refactor |  Create DAO UI/Validation

### PR DESCRIPTION
**Pull request by Da-Colon**
## Description
First I'd like to apologize for such a large PR. On top of the validation update using Yup. There was a design refactor as well. Didn't make sense to work in Yup into the old design or vice versa. not to mention the typing/import updates on a few files. I'll will comment through to ensure a easy read at least.

This PR covers the remove of the Context Provider that we were previous using to handle state, along with all the hooks and other components that were used for it. 

In their place is [Yup](https://www.npmjs.com/package/yup) for validation and [Formik](https://formik.org/) for form control
- Formik basically replaces the Context API provider and handles updating, setting, and validating input fields using its props and helper methods
- Yup is a JS validation libraries that allows you to create schemas that can used to validated values. `Formik` easily integrates with this libraries to validation any form.
- There are slight difference in designs and implementation. I need to get into the `Fractal-UI` and update the LabelWrapper and Input stylings to match the inputs shown. 
Chakra-UI also supports its use since it allows you to pass your own components as children and it works just the same.

## Notes
- Validation is taken out of the components and happens in one place `useCreateDAOScehma`
- No more Infinite requests in the background! Yups validation is asynchronous. so allows for us to check for valid ENS and resolve them easily. showing an error to user.
- Yup was a little tricker than I thought. I will answer any questions about the structure and setup. made an attempt to make reusable parts. as we create schema's in the future. I'm sure we can improve where needed.
- ParentTokenAllocationAmount is currently commented out. I may break into another PR As I need to create the new UI.

## Screenshots (if applicable)

localhost_3000_(Desktop) (34)

localhost_3000_(Desktop) (33)

localhost_3000_(Desktop) (35)

localhost_3000_(Desktop) (36)

localhost_3000_(Desktop) (37)

### CODE REVIEW

File: src/components/DaoCreator/refactor/BigNumberInput.tsx [hunk 1]
```diff
@@ -0,0 +1,160 @@
+import { FormControlOptions, Input, InputElementProps } from '@chakra-ui/react';
+import { utils, BigNumber, constants } from 'ethers';
+import { useState, useCallback, useEffect } from 'react';
+import { BigNumberValuePair } from '../types';
+
+export interface BigNumberInputProps
+  extends Omit<InputElementProps, 'value' | 'onChange'>,
+    FormControlOptions {
+  value: string;
```
Comment (Da-Colon) [inline src/components/DaoCreator/refactor/BigNumberInput.tsx:9]: Basically the change that effects the whole file. `BigNumber` -> `string`. No longer a `BigNumber` being passed in. But as there are other components using this file. I temporary created a copy of the input until the other forms have been updated.

File: src/components/DaoCreator/types.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+export interface CreatorFormState<T = BigNumberValuePair> {
+  essentials: DAOEssentials;
+  gnosis: GnosisConfiguration;
+  govToken: DAOGovenorToken<T>;
+  govModule: DAOGovenorModuleConfig<T>;
+  vetoGuard: DAOVetoGuardConfig<T>;
+  funding: DAOFunding;
+}
```
Comment (Da-Colon) [inline src/components/DaoCreator/types.ts:31]: By passing the type, we are able to updating the typing correctly for the purpose that we need it for. In this case. The DAO builder doesn't need to accept BigNumberValuePairs, keeping the typing clean while also allowing these same values with repearing types everywhere.

**Review by herbig — CHANGES_REQUESTED** [review]
- [ ] Can we add the breadcrumbs here (as in the designs)
- [x] (Minor) the Name field shows 1password icon, I had fixed this elswehere, so I can create a ticket to look into it here as well.
- [x] "Short description around governance and a link to learn more" was a placeholder I think, can we change that to "What type of governance would you like?"
- [x] Configure Voting Token - missing horizontal line between Token Supply and Initial Token Allocations
- [ ] Trash icon is misaligned if there is an error in allocations (screenshot)
Screen Shot 2023-02-13 at 8 42 34 PM
- [x] Adding a decimal place in the Token Supply field crashes the app
- [ ] Token Symbols don't have a limit on the smart contract, and there's no specific stipulation in the protocol, 3-5 is just a suggestion, so let's not limit that here
- [ ] leading / trailing spaces count towards the character limit, but the result is trimmed anyway, so if we keep this limit we should not count trailing leading space in the 5 char limit (also good to think about for other input fields)
- [ ] Unrelated to this PR, but we should change the Quorum description to something that says it must be YES votes..
- [x] Configure Multisig - Signature Threshold is set at 1 and cannot be changed for some reason
- [ ] Remove button is off center if there is an error, similar to trash icon on Voting allocations
- [ ] Should we use the trash icon here as well instead of Remove, to be consistent?
- [x] If I set Total Signers to 0, the single field disappears (which makes sense), but the Deploy button shifts up slightly.  Could we prevent that shift, or possibly not allow Total Signers to be 0 (default it to 1)?
- [ ] No padding between the horizontal divider and the Prev / Deploy buttons
- [ ] Trash icon should be disabled while deploying the DAO
- [ ] Clicking Trash icon during subDAO flow goes to app Homepage, rather than the parent DAO's homepage (or whereever they clicked from the manage DAO menu)
- [ ] subDAO flow final button should say "Create subDAO Proposal" rather than "Deploy" (this is the case on dev currently)
- [ ] Somehow I got into an error state on subDAO multisig flow: 
Error: invalid BigNumber string (argument="value", value="0.000000000000000016", code=INVALID_ARGUMENT, version=bignumber/5.7.0)
From this DAO (token voting parent): 
https://deploy-preview-966--fractal-framework-interface-dev.netlify.app/#/daos/0x8B7e4449FbbA6154C8A2De4E156D1A16e8fA7ada
Just after the Configure Multisig step, using the 1/1 default
- [ ] subDAO creation proposal with multisig parent, I select multisig -> enter signer address / Next -> From Configure Parent Controls click Prev -> I am now on the Token Voting Compose Governance step?

File: src/components/DaoCreator/StepController.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      );
+    case CreatorSteps.GNOSIS_GOVERNANCE:
+      return (
+        <GnosisMultisig
+          {...props}
+          step={step}
+          updateStep={updateStep}
+        />
+      );
+    // case CreatorSteps.FUNDING: {
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/components/DaoCreator/StepController.tsx:31]: yeah I'll go ahead and remove it here too

File: src/components/DaoCreator/formComponents/GnosisMultisig.tsx [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        <Box my={8}>
+          <LabelComponent
+            label={t('titleSignerAddresses')}
+            helper={t('subTitleSignerAddresses')}
+            isRequired={false}
+          >
+            <FieldArray name="gnosis.trustedAddresses">
+              {({ remove }) => (
+                <>
+                  {values.gnosis.trustedAddresses.map((trusteeAddress, i) => {
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/components/DaoCreator/formComponents/GnosisMultisig.tsx:85]: ha! good eye. I think before it was trustee. then made it a string[] and added adress. good call

File: src/components/DaoCreator/formComponents/UsulGovernance.tsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+import { Info } from '@decent-org/fractal-ui';
+import { useTranslation } from 'react-i18next';
+import { LabelComponent } from '../../ProposalCreate/InputComponent';
+import ContentBoxTitle from '../../ui/containers/ContentBox/ContentBoxTitle';
+import { StepButtons } from '../StepButtons';
+import { StepWrapper } from '../StepWrapper';
+import { BigNumberInput } from '../refactor/BigNumberInput';
+import { ICreationStepProps, CreatorSteps } from '../types';
+
+// @todo finish and deploy baby
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/components/DaoCreator/formComponents/UsulGovernance.tsx:14]: :facepalm:

File: src/components/DaoCreator/formComponents/UsulTokenAllocation.tsx [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      : null;
+  const amountErrorMessage =
+    indexedTokenAllocationError?.amount?.value && !tokenAllocation.amount.bigNumberValue.isZero()
+      ? indexedTokenAllocationError.amount.value
+      : null;
+  return (
+    <>
+      <LabelWrapper
+        errorMessage={addressErrorMessage}
+        key={index}
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/components/DaoCreator/formComponents/UsulTokenAllocation.tsx:42]: ah good catch leftover

File: src/components/DaoCreator/index.tsx [hunk 7]
```diff
@@ -1,27 +1,67 @@
...
+            case GovernanceTypes.GNOSIS_SAFE: {
+              const data = await prepareMultisigFormData({
+                ...values.essentials,
+                ...values.gnosis,
+                vetoGuard,
+              });
+              deployDAO(data);
+              return;
+            }
+            case GovernanceTypes.GNOSIS_SAFE_USUL: {
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/components/DaoCreator/index.tsx:40]: I agree, I resisted a naming refactor as much as possible, but a quick sweep though and nailed down a name for everything. @herbig

File: src/components/DaoCreator/formComponents/UsulTokenAllocations.tsx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+              >
+                {t('titleAddress')}
+              </Text>
+              <Text
+                textStyle="text-md-sans-regular"
+                color="grayscale.100"
+              >
+                {t('titleAmount')}
+              </Text>
+              {values.govToken.tokenAllocations.length > 1 && <Box>{/* EMPTY */}</Box>}
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/components/DaoCreator/formComponents/UsulTokenAllocations.tsx:78]: The empty spot here is to hold the grid correct order when there is an multiple allocations and the remove is added.

File: src/components/DaoCreator/formComponents/UsulTokenAllocation.tsx [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  tokenAllocation,
+  index,
+  remove,
+  errors,
+  values,
+  setFieldValue,
+}: ITokenAllocations) {
+  const { restrictChars } = useFormHelpers();
+  const indexedTokenAllocationError = (
+    errors?.govToken?.tokenAllocations as FormikErrors<
```
[the remark this answers is not in the document]
Comment (Da-Colon) [reply src/components/DaoCreator/formComponents/UsulTokenAllocation.tsx]: Check out the changes here and see if this is looking more like you were thinking? @mudrila

**Comment by Da-Colon** [comment]
> Somehow I got into an error state on subDAO multisig flow:
Error: invalid BigNumber string (argument="value", value="0.000000000000000016", code=INVALID_ARGUMENT, version=bignumber/5.7.0)
From this DAO (token voting parent):
https://deploy-preview-966--fractal-framework-interface-dev.netlify.app/#/daos/0x8B7e4449FbbA6154C8A2De4E156D1A16e8fA7ada
Just after the Configure Multisig step, using the 1/1 default

wasn't able to reproduce this error @herbig, let me know if you see this again

**Comment by herbig** [comment]
> > Somehow I got into an error state on subDAO multisig flow:
> > Error: invalid BigNumber string (argument="value", value="0.000000000000000016", code=INVALID_ARGUMENT, version=bignumber/5.7.0)
> > From this DAO (token voting parent):
> > https://deploy-preview-966--fractal-framework-interface-dev.netlify.app/#/daos/0x8B7e4449FbbA6154C8A2De4E156D1A16e8fA7ada
> > Just after the Configure Multisig step, using the 1/1 default
> 
> wasn't able to reproduce this error @herbig, let me know if you see this again

Hmm, no longer happening. Maybe some other refactor fixed it..

**Review by herbig — CHANGES_REQUESTED** [review]
- [x] Bug: set multisig Total Signers to 2, input an address in the first field, delete the second: Deploy button stays disabled
- [x] I don't like that you can't delete the 1 for Total Signers, I wonder if we could allow you to delete the whole field content and it would still show 1 field, and default to 1 when you leave the Total Signers input?
- [x] labelChooseGovernance and helperChooseGovernance don't render properly
- [ ] weren't we going to limit symbol input to 2-6 characters?
Screenshot 2023-02-17 at 6 43 58 PM

**Review by herbig — APPROVED** [review]
SHIP IT

**Pull request merged**